frontend/src/components/MonacoEditor/useMonaco.ts

```diff
@@ -171,11 +171,39 @@ const useMonaco = async (lang: string) => {
     })(),
   ]);
 
+  /**
+   * set new content in monaco editor
+   * use executeEdits API can preserve undo stack, allow user to undo/redo
+   * @param editorInstance Editor.IStandaloneCodeEditor
+   * @param content string
+   */
   const setContent = (
     editorInstance: Editor.IStandaloneCodeEditor,
     content: string
   ) => {
-    if (editorInstance) editorInstance.setValue(content);
+    if (editorInstance) {
+      const range = editorInstance?.getModel()?.getFullModelRange();
+      // get the current endLineNumber, or the default value is 100000, in case can not remove the current edits
```

Comment:
```suggestion
      // get the current endLineNumber, or use 100000 as the default
```

Reply:
I don't quite understand the "in case can not remove the current edits" part, I just removed that sentence